[Jia Xin] iP

jxt00 opened the pull request:
## Duke

> "Life is too complicated not to be orderly." – Martha Stewart ([source](https://dansilvestre.com/productivity-quotes/#10_Life_is_too_complicated_not_to_be_orderly_-_Martha_Stewart))

Duke makes your life simpler by organising your tasks for you. It's,

* text-based
* easy to learn
* ~~FAST~~ SUPER FAST to use
* **FREE** :+1::+1::+1:

All you need to do is,
1. download it from [here](https://github.com/jxt00/ip).
1. double-click it.
1. add your tasks.
1. let it manage your tasks for you :wink:

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public static void main(String[] args) {
    new Duke("data/tasks.txt").run();
}
```

Commit:
Add Gradle support

jxt00 committed:
Add basic functionalities to Duke.java

Implement a skeletal version of Duke that
* greets the user
* echos commands entered by the user
* exits when the user types bye

jxt00 committed:
Add add and list functionalities to Duke.java

jxt00 committed:
Add Task.java

jxt00 committed:
Add javadoc to Task.java

jxt00 committed:
Add the ability to mark/unmark tasks as done

Add javadoc to Duke.java

jxt00 committed:
Add Deadline.java, Event.java, Todo.java

* Deadline class: tasks that need to be done 
before a specific date/time
* Event class: tasks that start at a specific time and 
ends at a specific time
* Todo class: tasks without any date/time attached to it

Add javadoc to Deadline.java, Event.java, Todo.java

jxt00 committed:
Add support for tracking three types of tasks

* Deadline: tasks that need to be done 
before a specific date/time
* Event: tasks that start at a specific time and 
ends at a specific time
* Todo: tasks without any date/time attached to it

jxt00 committed:
Add support for tracking three types of tasks

jxt00 committed:
Add Automated Text UI Testing

* Remove whitespace in Todo's description
* Remove unnecessary newlines when printing

jxt00 committed:
Add Automated Text UI Testing

* Remove whitespace in Todo's description
* Remove unnecessary newlines when printing

jxt00 committed:
Merge branch 'master' of https://github.com/jxt00/ip

jxt00 committed:
Add DukeException.java and ExceptionType.java

DukeException handles
* empty input
* empty task description
* invalid dates for Deadline and Event

ExceptionType enum to represent different exception types
* empty input
* empty description
* invalid date

jxt00 committed:
Add exception handling

Duke handles the following errors:
* empty user input
* invalid command
* empty task description
* invalid date/time for Deadline and Event
* task to be marked/unmarked does not exist

jxt00 committed:
Add delete functionality to Duke.java

Replace tasks Array with ArrayList

jxt00 committed:
Remove ExceptionType enum

jxt00 committed:
DukeException class: remove unnecessary methods

jxt00 committed:
Add CommandType enum

jxt00 committed:
Add Parser.java to handle input parsing

Parser class: add methods to parse command and arguments

jxt00 committed:
Refactor Duke.java

Create a Duke class to run the chatbot. Add methods to handle greeting,
exiting, printing messages, as well as processing commands
so that codes will be neater.

jxt00 committed:
Add ErrorMessage.java to store error messages

jxt00 committed:
Add FileUtil to handle file operations

* write tasks to file
* read tasks from file

Task, Todo, Deadline, Event class: Add method to format tasks
for saving to file

Update test case and expected output

Duke class: Add initialiseTasks() method to initialize
list with saved tasks

jxt00 committed:
Task objects: store dates as a java.time.LocalDate

jxt00 committed:
Merge branch 'branch-Level-7'

jxt00 committed:
Merge branch 'branch-Level-8'

Fix conflict src/main/java/Parser.java

jxt00 committed:
Make the code more OOP

* Task class: add getter methods after changing attributes to private
* Task, Deadline, Event todo classes: remove
unnecessary formatForFile() method

* Add Ui class to deal with interactions with the user
* Add Storage class to deal with loading tasks from the file
and saving tasks in the file
* Add TaskList class to deal with add/delete tasks in the list

* Remove unnecessary FileUtil class

jxt00 committed:
Organize the classes into java packages

Extract more *Command classes that inherit from Command class:
* Add, Delete, Exit, List, Mark, Unmark classes

Add StorageException class to handle exceptions related to storage

Parser class: update parseCommand() method to return *Command object

jxt00 committed:
Remove unnecessary files

jxt00 committed:
Parser class: Fix parseDate() error

jxt00 committed:
Add JUnit Tests

Task, Deadline, Event classes: add equals() method

jxt00 committed:
Tweak the code to comply with the Coding Standard

jxt00 committed:
Add FindCommand class to find a task with a keyword

jxt00 committed:
Add JavaDoc comments

jxt00 committed:
Merge branch 'branch-A-CodingStandard'

jxt00 committed:
Merge branch 'branch-Level-9'

jxt00 committed:
Merge branch 'branch-A-JavaDoc'

jxt00 committed:
Standardize JavaDocs

Ui class: add printFoundTasks() method for FindCommand class

Task, Deadline, Event classes: add formatForFile() methods

Storage class: remove encodeTaskList(), encodeTaskToString(),
decodeTaskList(), decodeTaskFromString() methods

jxt00 committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-gradle

jxt00 committed:
Automate project builds using Gradle

jxt00 committed:
Merge branch 'branch-gradle'

jxt00 committed:
Add JavaFX

jxt00 committed:
Merge branch 'branch-Level-10'

jxt00 committed:
Use CheckStyle to fix some violations

Update Duke's GUI.

* Customize DialogBox to look like a text bubble.
* User picture is clipped to a circle.

jxt00 committed:
Merge branch 'branch-A-CheckStyle'

thikhinab commented on src/main/java/DialogBox.java:
```diff
@@ -0,0 +1,85 @@
+import java.io.IOException;
+import java.util.Collections;
+
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+import javafx.fxml.FXML;
+import javafx.fxml.FXMLLoader;
+import javafx.geometry.Pos;
+import javafx.scene.Node;
+import javafx.scene.control.Label;
+import javafx.scene.image.Image;
+import javafx.scene.image.ImageView;
+import javafx.scene.layout.HBox;
+import javafx.scene.shape.Circle;
```
Great work on explicitly listing all the classes imported!

thikhinab commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,29 @@
+/**
+ * Represents a command that can be executed.
+ */
+public abstract class Command {
```
Nice application of an abstract class!

thikhinab commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,26 @@
+
+    @Override
+    public String execute(TaskList tasks, Ui ui, Storage storage) throws DukeException {
+        Task t = tasks.getTask(index);
```
Would it be clearer if `t` was renamed as `task`?

jxt00 replied:
Hi, thanks for the feedback, I've renamed `t` as `deletedTask`.

thikhinab commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,112 @@
+            return new UnmarkCommand(Integer.parseInt(args[1]));
+            // Fallthrough
+        case "deadline":
+            if (args.length == 1) {
+                throw new DukeException(ErrorMessage.MESSAGE_UNKNOWN_DESC);
+            }
+            String[] deadline = args[1].split("/by", 2);
+            if (deadline.length == 1) {
+                throw new DukeException(ErrorMessage.MESSAGE_UNKNOWN_DATE);
+            }
+            return new AddCommand(new Deadline(deadline[0], parseDate(deadline[1])));
+            // Fallthrough
+        case "event":
+            if (args.length == 1) {
+                throw new DukeException(ErrorMessage.MESSAGE_UNKNOWN_DESC);
+            }
+            String[] event = args[1].split("/at", 2);
+            if (event.length == 1) {
+                throw new DukeException(ErrorMessage.MESSAGE_UNKNOWN_DATE);
+            }
+            return new AddCommand(new Event(event[0], parseDate(event[1])));
+            // Fallthrough
+        case "todo":
+            if (args.length == 1) {
+                throw new DukeException(ErrorMessage.MESSAGE_UNKNOWN_DESC);
+            }
+            return new AddCommand(new Todo(args[1]));
+            // Fallthrough
+        case "find":
+            if (args.length == 1) {
+                throw new DukeException(ErrorMessage.MESSAGE_UNKNOWN_KEYWORD);
+            }
+            return new FindCommand(args[1]);
+        case "bye":
+            return new ExitCommand();
+            // Fallthrough
+        default:
+            throw new DukeException(ErrorMessage.MESSAGE_UNKNOWN_COMMAND);
+        }
+    }
```
Would it be clearer if the method was broken down to smaller methods to be called at each case?

yusufaine replied:
I did it the same way as @jxt00 as well, but I think using switch-case makes it easier to tell that the entire block is solely trying to call the appropriate commands.

jxt00 replied:
Yep, as the method is longer than 30 LOC, I extracted out the parsing parts into smaller methods.

thikhinab commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,111 @@
+
+    private void createFile(Path filePath) throws StorageException {
+        try {
+            if (checkIfFileExists(filePath)) {
+                return;
+            }
+
+            Path parentDir = filePath.getParent();
+            if (!Files.exists(parentDir)) { // check if data folder exists
+                Files.createDirectories(parentDir);
+            }
+
+            Files.createFile(filePath);
```
Well done in making the path prominent!

thikhinab reviewed (commented):
Excellent work in coding style and quality! Only a few minor suggestions.

yusufaine commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,47 @@
+import duke.ui.Ui;
+import duke.util.TaskList;
+
+/**
+ * Represents a Personal Assistant Chatbot that helps a person to keep track of various tasks.
+ */
```
Good job on this, I wasn't so on what comment to write for this but you gave me a really good idea 👍

yusufaine commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,29 @@
+    /**
+     * Returns false if the command is not an ExitCommand.
+     */
+    public boolean isExit() {
+        return false;
+    }
```
Good use on this 👍 ! I did something similar but naively made it an abstract method which meant that I had to implement it in all my subclasses 😓

yusufaine commented on src/main/java/duke/exception/ErrorMessage.java:
```diff
@@ -0,0 +1,18 @@
+/**
+ * Stores error messages to be shown to the user.
+ */
+public class ErrorMessage {
```
Would this be better if it was an Enum?

jxt00 replied:
Hi, you can take at look at adressbook-level2 [Messages.java](https://github.com/se-edu/addressbook-level2/blob/master/src/seedu/addressbook/common/Messages.java) which I referred to.

yusufaine commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,112 @@
+     */
+    public static Command parseCommand(String input) throws DukeException {
+        if (input.isBlank()) {
+            throw new DukeException(ErrorMessage.MESSAGE_INVALID_INPUT);
```
Good use of the `ErrorMessage` class! Now I'm considering implementing that in mine as well 😄

yusufaine commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,111 @@
+    public static final String DEFAULT_FILEPATH = "duke.txt";
+    private final Path filePath;
+
+    public Storage() {
+        this(DEFAULT_FILEPATH);
+    }
+
+    /**
+     * Creates a new storage with the specified file path.
+     *
+     * @param filePath Path of the file containing the stored tasks.
+     */
+    public Storage(String filePath) {
+        this.filePath = Paths.get(filePath);
+    }
```
Good implementation! 

I did not consider checking for default/allowing users to decide where they want their data file to be stores,

yusufaine commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,68 @@
+                + this.by.format(DateTimeFormatter.ofPattern("MMM d yyyy")) + ")";
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        if (!super.equals(o)) {
+            return false;
+        }
+
+        Deadline deadline = (Deadline) o;
+        return by.equals(deadline.by);
+    }
```
Not sure if this is good practice per-se, I remembered getting grilled that we should not be using `getClass()` 2030.

jxt00 replied:
I actually just used IntelliJ to generate it for me :sweat_smile: Might remove it after I clean up the JUnit tests.

yusufaine commented on src/main/java/duke/util/TaskList.java:
```diff
@@ -0,0 +1,112 @@
+     * @throws DukeException if the index is invalid.
+     */
+    public Task getTask(int index) throws DukeException {
+        int zeroBasedIndex = index - 1;
+        if (zeroBasedIndex < 0 || zeroBasedIndex >= tasks.size()) {
+            throw new DukeException(ErrorMessage.MESSAGE_INVALID_INDEX);
+        }
```
Good job in considering these errors and explicitly naming the variable `zeroBasedIndex` to avoid confusion!

yusufaine reviewed (commented):
Overall good code quality and readability. This gave me some ideas on how to modify some things in my own implementation!

jxt00 committed:
Modify getTasksWithKeyword() method to use regex

Method loops through the list and adds tasks containing the
keyword, even if it's just a substring, to a new list.

Regex adds only the task containing the whole keyword.

Let's,
* update the algorithm to use match whole words only
* use Streams

jxt00 committed:
Main class: set title and icon for customization

Task, Event, Deadline classes: remove unnecessary getters

jxt00 committed:
Modify *Command and Ui classes to use string format

String appending is long and messy.

Use String.format() which is neater and easier to read.
Store messages as constant for easier updating in the future.

jxt00 committed:
Parser class: Apply SLAP to parseCommand() method

parseCommand() is too long (exceeds 30 LOC) and has common
behaviors. The common behaviors across cases cause code duplication.

Extracting the common behavior into methods that allows us to reuse
them when dealing with the different commands. For example, validity
checking of the index can be done in a method and called repeatedly
for mark, unmark and delete commands.

Let's put the various validity checks into new, smaller methods
that can be called by different cases.

jxt00 committed:
TaskList class: extract a common index conversion method

getTask() and deleteTask() have some common behavior (converting
one-based index to zero-based).

Extracting the common behavior into method that allows us to reuse
them. For example, both getTask() and deleteTask() can now call
convertIndexToZeroBased().

jxt00 committed:
Merge pull request #1 from jxt00/branch-A-Assertions

TaskList class: extract a common index conversion method

jxt00 committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/command/DeleteCommand.java
#	src/main/java/duke/parser/Parser.java
#	src/main/java/duke/util/TaskList.java

jxt00 committed:
Merge pull request #2 from jxt00/branch-A-CodeQuality

Request to merge branch-A-CodeQuality

jxt00 committed:
Merge branch 'master' into branch-A-Streams

# Conflicts:
#	src/main/java/duke/util/TaskList.java

jxt00 committed:
Merge pull request #3 from jxt00/branch-A-Streams

Modify getTasksWithKeyword() method to use regex

jxt00 committed:
Add HelpCommand

Task, Deadline, Event classes: remove unnecessary equals() method

Update JUnit tests

jxt00 committed:
Add ScheduleCommand

Set up CI with gradle.yml

Task, Deadline, Event classes: add getDate() method

TaskList class: add method that filters for tasks that
occurs on a specified date.

Ui class: update showTasks() method to display tasks that
occurs on a specified date.

Parser class: add method to create new ScheduleCommand

jxt00 committed:
Merge branch 'branch-BCD-Extension'

jxt00 committed:
Create gradle.yml

jxt00 committed:
Fix build.gradle failure on GitHub

jxt00 committed:
Finalize features and UI

jxt00 committed:
Create _config.yml

jxt00 committed:
Fix header comment to follow coding standard

jxt00 committed:
Update build.gradle checkstyle version to 8.32

jxt00 committed:
Fix checkstyle errors in test classes

jxt00 committed:
Update app to close upon "bye" input

jxt00 committed:
Fix minor mistake in README.md
